File: TD/TD1/exo3/exo3.cpp
#include <string>
#include <iostream>

class Vehicule {
private:
	std::string modele;
	int anneAchat;
	int prixAchat;
	std::string immatriculation;
	char permis;

public:
	Vehicule(std::string m,int y,int p,std::string i, char l) {
		modele = m;
		anneAchat = y;
		prixAchat = p;
		immatriculation = i;
		permis = l;

	}

	int age(){
		return 2019-anneAchat;
	}

	int coutLocation() {
		if(age()<1) return prixAchat/200;
		else return prixAchat/250;
	}

	void afficher() {
		std::cout<<modele<<" "<<immatriculation<<std::endl;
	}
};

class Voiture : public Vehicule {
private:
	bool autoradio;

public:
	Voiture(std::string m,int y,int p,std::string i, char l, bool r) : 
	Vehicule(m,y,p,i,l), autoradio(r){
	}

	void ADDautoradio () {
		autoradio = true;
	}

	void RMautoradio () {
		this->autoradio = false;
	}
};

class Utilitaire : public Vehicule {
private:
	float volume;

public:
	Utilitaire(std::string m,int y,int p,std::string i, char l, float v) : 
	Vehicule(m,y,p,i,l){
		volume = v;
	}

	bool peutTransporterVolume (int v) {
		return (v<=volume);
	}
};

class AutoCar : public Utilitaire {
private:
	int voyageurs;

public:
	AutoCar(std::string m,int y,int p,std::string i, char l, float v, int t):
	Utilitaire(m,y,p,i,l,v){
		voyageurs = t;
	}

	bool peutTransporterPassagers(int p, int v){
		return (p<=voyageurs && peutTransporterVolume(p*v));
	}
};



int main () {
	Voiture V1 = new Voiture("Twingo",2005,10000,"1234 TZ 49",'B',true);
	V1.afficher();

	Utilitaire CA1 = new Utilitaire("J9",2018-5,20000,"2345 FA 49",'B',5);
	//std::cout<<(CA1.coutLocation()+" "+CA1.peutTransporterVolume(4));

	//std::cout<<(V1.coutLocation());

	AutoCar AU1 = new AutoCar("FR1",2018-2,"3456 IJ 49",'D',3.0,53);
	//std::cout<<(AU1.peutTransporterPassagers(40,0.1));
}

File: TD/TD1/exo3/exo3.py
class Vehicule:
	def __init__(self,m,y,c,i,l):
		self.modele=m
		self.annee=y
		self.prix=c
		self.immatriculation=i
		self.permis=l

	def age(self):
		return 2019-self.annee

	def coutLocation(self):
		if(age()<1): return self.prix/200
		else: return self.prix/250

	def __str__(self):
		return self.modele +" "+ self.immatriculation

class Voiture(Vehicule):
	def __init__(self,m,y,c,i,l,r):
		Vehicule.__init__(self,m,y,c,i,l)
		self.autoradio = r

	def ADDautoradio(self):
		self.autoradio = True

	def RMautoradio(self):
		self.autoradio = False

class Utilitaire(Vehicule):
	def __init__(self,m,y,c,i,l,v):
		Vehicule.__init__(self,m,y,c,i,l)
		self.volume = v

		def peutTransporterVolume(self,v):
			return v<=self.volume

class AutoCar(Utilitaire):
	def __init__(self,m,y,c,i,l,v,p):
		Utilitaire.__init__(self,m,y,c,i,l,v)
		self.passagers = p

	def peutTransporterPassagers(self,p,v):
		return p<=self.passagers and super().peutTransporterVolume(p*v)



car = Voiture("Twingo",2005,10000,"1234 TZ 49",'B',True)
print(car)
car.RMautoradio()

van = Utilitaire("J9",2018-5,20000,"2345 FA 49",'B',5)
print(van)
a= van.peutTransporterVolume()

bus = AutoCar("FR1",2018-2,90000,"3456 IJ 49",'D',3.0,53)
b= bus.peutTransporterPassagers(40,0.1)